torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -353,6 +353,105 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
   return {std::move(payload), std::move(tensors)};
 }
 
+TensorPipeEntry tensorpipeSerialize(const Message& rpcMessage) {
+  tensorpipe::Message tpMessage;
+  std::vector<torch::Tensor> reservedTensors;
+  std::vector<std::vector<uint8_t>> copiedTensors;
+
+  const std::vector<char>& payload = rpcMessage.payload();
+  c10::List<at::Tensor> tensors = cloneSparseTensors(rpcMessage.tensors());
+
+  // Payload
+  tpMessage.data = (uint8_t*)(payload.data());
+  tpMessage.length = payload.size();
+
+  // Metadata - encode rpc message type + message id
+  tpMessage.metadata = c10::str(
+      static_cast<int>(rpcMessage.type()),
+      ":",
+      static_cast<int>(rpcMessage.id()));
+
+  // Tensors
+  tpMessage.tensors.reserve(tensors.size());
+  for (const at::Tensor& tensor : tensors) {
+    // Keep original user tensors and cloned sparse tensors
+    reservedTensors.push_back(tensor);
+    tensorpipe::Message::Tensor tpTensor;
+    torch::jit::Pickler pickler([&](const void* buf, size_t sz) -> size_t {
+      tpTensor.metadata.append(static_cast<const char*>(buf), sz);
+      return sz;
+    });
+    pickler.protocol();
+    pickler.pushIValue(tensor);
+    pickler.stop();
+    const auto& tensorDataVec = pickler.tensorData();
+    TORCH_INTERNAL_ASSERT(
+        tensorDataVec.size() == 1, "There should be single pickled tensor");
+    const auto& tensorData = tensorDataVec.front();
+    // Enforce memory copy if tensor is created from torch::from_blob, means
+    // that the tensor doesn't own the memory.
+    if (!tensorData.storageHasDeleter()) {
+      std::vector<uint8_t> storageData(tensorData.sizeInBytes());
+      memcpy(storageData.data(), tensorData.data(), storageData.size());
+      tpTensor.data = storageData.data();
+      copiedTensors.push_back(std::move(storageData));
+    } else {
+      tpTensor.data = (uint8_t*)(tensorData.data());
+    }
+    tpTensor.length = tensorData.sizeInBytes();
+    tpMessage.tensors.push_back(std::move(tpTensor));
```

Comment:
Would this work for tensors that are non-contiguous?

Reply:
None of the RpcAgents support sparse/non-contiguous tensors. I believe it was decided late last year to defer that work to the Pickler, which would make it work in many more situations as well.

My impression is that it's not currently high-pri in Pickler, though probably doable if there's a motivating use case.

My suggestion would be to get v1.0 in (and e.g. building on OSS/github) before we worry about cases like this.

Reply:
I thought the  pickler does handle non-contiguous (dense non-contiguous tensors not referring to sparse tensors) tensors by storing the strides appropriately: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/serialization/pickler.cpp#L373. I did add a simple test as well to check that this seems to work for our current RPC agents: https://github.com/pytorch/pytorch/pull/36705 (and since this logic is similar, it probably would work here as well?)